src/utils/src/ExponentialBackoffUtils.c

```diff
@@ -0,0 +1,223 @@
+    CHK(pExponentialBackoffConfig != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    *pExponentialBackoffConfig = *pBackoffConfig;
+
+    pExponentialBackoffState->pExponentialBackoffConfig = pExponentialBackoffConfig;
+    CHK_STATUS(resetRetryState(pExponentialBackoffState));
+
+    CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppBackoffState) = pExponentialBackoffState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffState);
+        SAFE_MEMFREE(pExponentialBackoffConfig);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+UINT64 getRandomJitter(UINT32 jitterFactor) {
+    return (RAND() % jitterFactor) * HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+}
+
+UINT64 power(UINT32 base, UINT32 exponent) {
+    UINT64 result = 1;
+    while (exponent-- > 0) {
+        result *= base;
+    }
+    return result;
+}
+
+UINT64 calculateWaitTime(PExponentialBackoffState pRetryState, PExponentialBackoffConfig pRetryConfig) {
+    return power(DEFAULT_KVS_EXPONENTIAL_FACTOR, pRetryState->currentRetryCount) * pRetryConfig->retryFactorTime;
+}
+
+STATUS validateAndUpdateExponentialBackoffStatus(PExponentialBackoffState pExponentialBackoffState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    switch (pExponentialBackoffState->status) {
+        case BACKOFF_NOT_STARTED:
+            pExponentialBackoffState->status = BACKOFF_IN_PROGRESS;
+            DLOGD("Status changed from BACKOFF_NOT_STARTED to BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_IN_PROGRESS:
+            // Do nothing
+            DLOGD("Current status is BACKOFF_IN_PROGRESS");
+            break;
+        case BACKOFF_TERMINATED:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Current status is BACKOFF_TERMINATED");
+            retStatus = STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE;
+            break;
+        default:
+            DLOGD("Cannot execute exponentialBackoffBlockingWait. Unexpected state [%"PRIu64"]", pExponentialBackoffState->status);
+            retStatus = STATUS_EXPONENTIAL_BACKOFF_INVALID_STATE;
+    }
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffBlockingWait(PExponentialBackoffState pRetryState) {
+    ENTERS();
+    PExponentialBackoffConfig pRetryConfig = NULL;
+    UINT64 currentTime = 0, waitTime = 0, jitter = 0;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pRetryState != NULL, STATUS_NULL_ARG);
+    CHK(pRetryState->pExponentialBackoffConfig != NULL, STATUS_NULL_ARG);
+
+    CHK_STATUS(validateAndUpdateExponentialBackoffStatus(pRetryState));
+
+    pRetryConfig = pRetryState->pExponentialBackoffConfig;
+
+    // If retries is exhausted, return error to the application
+    if (pRetryConfig->maxRetryCount != KVS_INFINITE_EXPONENTIAL_RETRIES) {
+        CHK(!(pRetryConfig->maxRetryCount == pRetryState->currentRetryCount),
+            STATUS_EXPONENTIAL_BACKOFF_RETRIES_EXHAUSTED);
+    }
```

Comment:
This might be the best to be expressed as a single line

CHK(maxRetryCount == KVS_INFINITE_EXP || maxRetry > currentRetry, STATUS_EXP_BACKOFF_EXHAUSTED);

Reply:
I had a hard time writing this line of code with CHK. I always get confused with the `!` in CHK macro. Although what you're suggesting is one liner, I find the if condition more intuitive to read.

Reply:
It might take a little while to get used to these macros but believe me that creating those in the first place and meticulously using those allows the entire codebase to avoid a class of issues that normally plague native codebases.

In this case, you can certainly use the first if to avoid the || but try to not check the negation - that is really hard to read.

Reply:
+1 I think this change would be cleaner

Reply:
Cool. Doing it.